Add CheckDiff hooks

gnustavo opened the pull request.

gnustavo committed:
Add CheckDiff hooks

mikkoi commented:
A very good addition to the hooks.

marceloavilaoliveira commented on script/find-secret-leakage-in-git-diff.pl:
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env perl
```
Where is the call to find-secret-leakage-in-git-diff.pl script?

gnustavo replied:
It must be configured as in the example at line 242 of the file CheckDiff.pm.

marceloavilaoliveira replied:
Got it.

marceloavilaoliveira commented on README.pod:
```diff
@@ -49,6 +49,8 @@ plugins provided by the distribution are these:
 
 =item * L<Git::Hooks::CheckCommit> - enforce commit policies
 
+=item * L<Git::Hooks::CheckDiff> - check differences between commits
```
Why "check differences between commits"? If I understood corretly If there's a line with some issue but it wasn't change then Git::Hooks will not alert. Is it right?

gnustavo replied:
Yes. The idea is that this hook will generate a git-diff and pass its output to one or more commands to check it. It can be used in a pre-commit or in a pre-receive hook, for example. In a pre-commit it uses git-diff-index to generate the next commit diff. In a pre-receive it uses a git-diff-tree to generate the next push diff.

How could I make this clearer in a single line?

marceloavilaoliveira replied:
I think the message is good enough, don't worry about it. I was a little confused because the spaces verification happens in the whole file and not just in the changed/added lines, I thought this new feature would work the same way. Actually, it would be good if the user could choose if he/she wants to verify the whole file or just the changed/added lines. What do you think?

gnustavo replied:
There is the Git::Hooks::CheckFile hook for checking the whole files. CheckDiff is meant to be more efficient and more specific when the check need only see the changed lines.

marceloavilaoliveira replied:
Uhm... but we can't use "find-secret-leakage-in-git-diff.pl" in the CheckFile, can we?

gnustavo replied:
Nope. It's meant to check a diff. The idea is that you use it to check only newly added parts in your code to avoid introducing new leakages. I don't think CheckFile would be a good fit for finding secret leakages. If you want to check all of your code I think you should process all of your code at once, using something like SonarQube.

Anyway, this is such a small script that it would be easy to create another one called find-secret-leakage.pl that would check all the lines in the input, not bothering to parse a diff output. I wouldn't use it in a CheckFile, but it could be useful to check all of your files in a script.

marceloavilaoliveira replied:
Ok

gnustavo committed:
Subsumes private keys tokens in a general regex

The private keys all have a very similar tokens. It's simpler and more
efficient to use a single token with a more general regex to match them
all.

gnustavo committed:
Allows one to mark pseudo-secrets

Sometimes a secret isn't dangerous. It can be used only on testing
environments or as an example in a document. There must be a way to mark
these pseudo-secrets so that the find-secret-leakage-in-git-diff.pl does
not denounce them.

Now there is.
